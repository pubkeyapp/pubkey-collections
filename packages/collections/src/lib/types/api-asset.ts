export interface ApiAsset {
  id: string
  content: {
    metadata: {
      name: string
      description: string
      attributes: Array<{
        trait_type: string
        value: string
      }>
    }
    files: {
      uri: string
      cdn_uri: string
      mime: string
    }
  }
  grouping: Array<{
    group_type: string
    group_value: string
  }>
}
