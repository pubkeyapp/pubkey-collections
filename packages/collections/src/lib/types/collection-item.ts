import { CollectionTrait } from './collection-trait'

export interface CollectionItem {
  id: string
  name: string
  description?: string
  image?: string
  collectionId: string
  traits: CollectionTrait[]
}
