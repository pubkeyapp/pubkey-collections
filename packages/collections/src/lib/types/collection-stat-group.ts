import { CollectionStat } from './collection-stat'

export interface CollectionStatGroup {
  key: string
  items: CollectionStat[]
}
