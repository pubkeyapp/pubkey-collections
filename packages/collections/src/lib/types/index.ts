export * from './api-asset'
export * from './collection'
export * from './collection-combo'
export * from './collection-item'
export * from './collection-set'
export * from './collection-trait'
export * from './collection-trait-map'
export * from './collection-trait-summary'
