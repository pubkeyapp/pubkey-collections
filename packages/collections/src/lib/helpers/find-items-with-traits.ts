import { CollectionItem, CollectionTrait } from '../types'

export function findItemsWithTraits(items: CollectionItem[], traits: CollectionTrait[]): CollectionItem[] {
  const result: CollectionItem[] = []

  for (const item of items) {
    const hasAllTraits = traits.every((trait) =>
      item.traits.find((t) => t.key === trait.key && t.value === trait.value),
    )

    if (hasAllTraits) {
      result.push(item)
    }
  }

  return result
}
