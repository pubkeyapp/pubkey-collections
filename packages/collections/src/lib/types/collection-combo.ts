import { CollectionItem } from './collection-item'
import { CollectionTrait } from './collection-trait'

export interface CollectionCombo {
  name: string
  sortKey?: string
  traits: CollectionTrait[]
  items?: CollectionItem[]
}
