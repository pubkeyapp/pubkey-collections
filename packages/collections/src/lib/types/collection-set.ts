import { Collection } from './collection'
import { CollectionItem } from './collection-item'
import { CollectionTraitMap } from './collection-trait-map'

export interface CollectionSet {
  collection: Collection
  items: CollectionItem[]
  traits: CollectionTraitMap
}
