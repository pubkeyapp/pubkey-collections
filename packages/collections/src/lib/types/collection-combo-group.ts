import { CollectionCombo } from './collection-combo'

export interface CollectionComboGroup {
  name: string
  combos: CollectionCombo[]
}
