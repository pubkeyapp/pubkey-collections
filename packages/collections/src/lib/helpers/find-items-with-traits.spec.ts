import { CollectionItem, CollectionTrait } from '../types'
import { findItemsWithTraits } from './find-items-with-traits'

describe('find-items-with-traits', () => {
  it('should find items with traits', () => {
    // Arrange
    const items: CollectionItem[] = [
      {
        id: 'item-1',
        name: 'Item 1',
        traits: [
          { key: 'Background', value: 'Red' },
          { key: 'Foreground', value: 'Green' },
        ],
        collectionId: 'items',
        image: 'image-1',
      },
      {
        id: 'item-2',
        name: 'Item 2',
        traits: [
          { key: 'Background', value: 'Blue' },
          { key: 'Foreground', value: 'Yellow' },
        ],
        collectionId: 'items',
        image: 'image-2',
      },
      {
        id: 'item-3',
        name: 'Item 3',
        traits: [
          { key: 'Background', value: 'Red' },
          { key: 'Foreground', value: 'White' },
          { key: 'Shape', value: 'Square' },
        ],
        collectionId: 'items',
        image: 'image-3',
      },
      {
        id: 'item-4',
        name: 'Item 4',
        traits: [
          { key: 'Background', value: 'Red' },
          { key: 'Foreground', value: 'White' },
          { key: 'Shape', value: 'Circle' },
        ],
        collectionId: 'items',
        image: 'image-4',
      },
    ]
    const traits: CollectionTrait[] = [
      { key: 'Background', value: 'Red' },
      { key: 'Foreground', value: 'White' },
    ]

    // Act
    const result = findItemsWithTraits(items, traits)

    // Assert
    expect(result).toMatchSnapshot()
  })
})
