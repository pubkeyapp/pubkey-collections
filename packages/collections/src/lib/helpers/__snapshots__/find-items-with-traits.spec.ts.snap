// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find-items-with-traits should find items with traits 1`] = `
[
  {
    "collectionId": "items",
    "id": "item-3",
    "image": "image-3",
    "name": "Item 3",
    "traits": [
      {
        "key": "Background",
        "value": "Red",
      },
      {
        "key": "Foreground",
        "value": "White",
      },
      {
        "key": "Shape",
        "value": "Square",
      },
    ],
  },
  {
    "collectionId": "items",
    "id": "item-4",
    "image": "image-4",
    "name": "Item 4",
    "traits": [
      {
        "key": "Background",
        "value": "Red",
      },
      {
        "key": "Foreground",
        "value": "White",
      },
      {
        "key": "Shape",
        "value": "Circle",
      },
    ],
  },
]
`;
